main_test.go

```diff
@@ -0,0 +1,224 @@
+	return nil, nil
+}
+
+func (m *MockDeployment) UpdateStatus(*extensions.Deployment) (*extensions.Deployment, error) {
+	return nil, nil
+}
+
+func (m *MockDeployment) Watch(opts api.ListOptions) (watch.Interface, error) {
+	return nil, nil
+}
+
+func (m *MockDeployment) Rollback(*extensions.DeploymentRollback) error {
+	return nil
+}
+
+func (m *MockKubeClient) Deployments(namespace string) kclient.DeploymentInterface {
+	return &MockDeployment{
+		client: m,
+	}
+}
+
+func NewMockKubeClient() *MockKubeClient {
+	return &MockKubeClient{
+		Deployment: &extensions.Deployment{
+			Spec: extensions.DeploymentSpec{
+				Replicas: 3,
+			},
+		},
+	}
+}
+
+func NewMockPodAutoScaler(kubernetesDeploymentName string, kubernetesNamespace string, max int, min int) *scale.PodAutoScaler {
+	mockClient := NewMockKubeClient()
+
+	return &scale.PodAutoScaler{
+		Client:     mockClient,
+		Min:        min,
+		Max:        max,
+		Deployment: kubernetesDeploymentName,
+	}
+}
+
+type MockSQS struct {
+	QueueAttributes *sqs.GetQueueAttributesOutput
+}
+
+func (m *MockSQS) GetQueueAttributes(*sqs.GetQueueAttributesInput) (*sqs.GetQueueAttributesOutput, error) {
+	return m.QueueAttributes, nil
+}
+
+func (m *MockSQS) SetQueueAttributes(input *sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error) {
+	m.QueueAttributes = &sqs.GetQueueAttributesOutput{
+		Attributes: input.Attributes,
+	}
+	return &sqs.SetQueueAttributesOutput{}, nil
+}
+
+func NewMockSqsClient() *mainsqs.SqsClient {
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("50")
+
+	return &mainsqs.SqsClient{
+		Client: &MockSQS{
+			QueueAttributes: &sqs.GetQueueAttributesOutput{
+				Attributes: Attributes,
+			},
+		},
+		QueueUrl: "sqs.test.com",
+	}
+}
+
+func TestRunMaxMin(t *testing.T) {
+	// override default vars for testing
+	*pollInterval = 1 * time.Second
+	*scaleDownCoolPeriod = 1 * time.Second
+	*scaleUpCoolPeriod = 1 * time.Second
+	*scaleUpMessages = 100
+	*scaleDownMessages = 10
+	*maxPods = 5
+	*minPods = 1
+	*awsRegion = "us-east-1"
+
+	*sqsQueueUrl = "sqs.test.com"
+	*kubernetesDeploymentName = "test"
+	*kubernetesNamespace = "test"
+
+	p := NewMockPodAutoScaler(*kubernetesDeploymentName, *kubernetesNamespace, *maxPods, *minPods)
+	s := NewMockSqsClient()
+
+	go Run(p, s)
+
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("100")
+	input := &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	time.Sleep(10 * time.Second)
+	deployment, _ := p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(*maxPods), deployment.Spec.Replicas)
+
+	Attributes = make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("10")
+	input = &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	time.Sleep(10 * time.Second)
```

Comment:
Testing the Run function is more of an integration test. I agree that sleeping is a bit of a code smell. Any suggestions on alternatives?